### PR TITLE
Cache MX and A server lookups

### PR DESCRIPTION
**Pull request by ianbayne**
Closes https://github.com/micke/valid_email2/issues/179.

### CODE REVIEW

**Comment by ianbayne** [comment]
Hi @micke, please let me know if this isn't what you had in mind. Happy to follow a different path 👍

**Comment by micke** [comment]
This is a good start, i'm however thinking that we might need to consider two factors still:

1. The cache growing to be too big.
2. The cache going stale, i think it would be really important for us to take the DNS records TTL into consideration to make sure that we don't cache the MX servers for too long.

**Comment by ianbayne** [comment]
Thanks for the feedback, @micke. Any thoughts on a maximum size for the cache?

**Comment by ianbayne** [comment]
> This is a good start, i'm however thinking that we might need to consider two factors still:
> 
>     1. The cache growing to be too big.
> 
>     2. The cache going stale, i think it would be really important for us to take the DNS records TTL into consideration to make sure that we don't cache the MX servers for too long.

@micke, I've gone ahead and pushed up some additional commits that should meet these requirements. I've arbitrarily selected 1,000 as the max cache limit. Please feel free to suggest a different value! Let me know what you think, please 🙇

**Comment by micke** [comment]
@ianbayne Great work! What do you think about maybe moving the caching logic to it's own class?

**Comment by ianbayne** [comment]
> @ianbayne Great work! What do you think about maybe moving the caching logic to it's own class?

Thanks! Sure, works for me. I’ll get on it when I’ve next got some spare time 👍

File: spec/spec_helper.rb [hunk 1]
```diff
@@ -5,6 +5,7 @@
 require 'rspec-benchmark'
 RSpec.configure do |config|
   config.include RSpec::Benchmark::Matchers
+  config.default_formatter = 'doc'
```
Comment (ianbayne) [inline spec/spec_helper.rb:8]: I prefer this personally, but feel free to delete if you'd rather not have it.

File: lib/valid_email2/dns_records_cache.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
+module ValidEmail2
+  class DnsRecordsCache
```
Comment (ianbayne) [inline lib/valid_email2/dns_records_cache.rb:2]: > `DnsRecordsCache`

Definitely open to different/better naming for this!

**Comment by ianbayne** [comment]
> @ianbayne Great work! What do you think about maybe moving the caching logic to it's own class?

This should be ready for another review, @micke. Lots of stubbing going on in the tests, but hopefully not an excessive amount.

**Comment by mscrivo** [comment]
This is great stuff @ianbayne!  One other optimization I can see being useful here is to move the setup of Resolv::DNS::Config into these functions where it's actually used. We're seeing a perf problem with heavy use of this gem where it's spending a lot of time setting up this config even though we're only calling `valid?` and never actually using it.

**Comment by ianbayne** [comment]
Sure @mscrivo, happy to make that change. See commit https://github.com/micke/valid_email2/pull/256/commits/eb794f223d099376bac12a77ec9a5459f1989a59.

@micke, please let me know if you'd prefer this commit to be part of a separate PR.

**Comment by micke** [comment]
@ianbayne huge thank you for this! Great work!

**Comment by ianbayne** [comment]
> @ianbayne huge thank you for this! Great work!

Thanks, @micke! Was a fun feature and I learned a lot.

**Comment by natematykiewicz** [comment]
I found this gem today and wanted to know how DNS caching was handled. I started reading, and realized that DNS records aren't actually cached.

They're cached into DNS record cache instances, that are instantiated within `ValidEmail2::Address` instances [here](https://github.com/micke/valid_email2/blob/d60b89b9f08de11a168cf951f901a2f32e2c4b32/lib/valid_email2/address.rb#L159) and [here](https://github.com/micke/valid_email2/blob/d60b89b9f08de11a168cf951f901a2f32e2c4b32/lib/valid_email2/address.rb#L174).

But the problem is, that only caches subsequent DNS calls on the same `ValidEmail2::Address` instance, which never happens. The gem instantiates a new `ValidEmail2::Address` instance [every time the validator runs](https://github.com/micke/valid_email2/blob/d60b89b9f08de11a168cf951f901a2f32e2c4b32/lib/valid_email2/email_validator.rb#L15).

So in practice, nothing is actually cached.

```ruby
address1 = ValidEmail2::Address.new('a@gmail.com')
=>
#<ValidEmail2::Address:0x00000001668d6550


address2 = ValidEmail2::Address.new('a@gmail.com')
=>
#<ValidEmail2::Address:0x0000000166f39a28


address1.send(:mx_or_a_servers)
=>
[#<Resolv::DNS::Resource::IN::MX:0x00000001669f6a70
  @exchange=
   #<Resolv::DNS::Name: alt2.gmail-smtp-in.l.google.com.>,
  @preference=20,
  @ttl=240>,
 #<Resolv::DNS::Resource::IN::MX:0x00000001669f60c0
  @exchange=
   #<Resolv::DNS::Name: alt3.gmail-smtp-in.l.google.com.>,
  @preference=30,
  @ttl=240>,
 #<Resolv::DNS::Resource::IN::MX:0x00000001669f5710
  @exchange=
   #<Resolv::DNS::Name: alt4.gmail-smtp-in.l.google.com.>,
  @preference=40,
  @ttl=240>,
 #<Resolv::DNS::Resource::IN::MX:0x00000001669f4e28
  @exchange=#<Resolv::DNS::Name: gmail-smtp-in.l.google.com.>,
  @preference=5,
  @ttl=240>,
 #<Resolv::DNS::Resource::IN::MX:0x00000001669f4478
  @exchange=
   #<Resolv::DNS::Name: alt1.gmail-smtp-in.l.google.com.>,
  @preference=10,
  @ttl=240>]


address2.send(:mx_or_a_servers)
=>
[#<Resolv::DNS::Resource::IN::MX:0x0000000163e7c110
  @exchange=#<Resolv::DNS::Name: gmail-smtp-in.l.google.com.>,
  @preference=5,
  @ttl=943>,
 #<Resolv::DNS::Resource::IN::MX:0x0000000163e7b4e0
  @exchange=
   #<Resolv::DNS::Name: alt1.gmail-smtp-in.l.google.com.>,
  @preference=10,
  @ttl=943>,
 #<Resolv::DNS::Resource::IN::MX:0x0000000163e7a6f8
  @exchange=
   #<Resolv::DNS::Name: alt3.gmail-smtp-in.l.google.com.>,
  @preference=30,
  @ttl=943>,
 #<Resolv::DNS::Resource::IN::MX:0x0000000163e79af0
  @exchange=
   #<Resolv::DNS::Name: alt2.gmail-smtp-in.l.google.com.>,
  @preference=20,
  @ttl=943>,
 #<Resolv::DNS::Resource::IN::MX:0x0000000163e78998
  @exchange=
   #<Resolv::DNS::Name: alt4.gmail-smtp-in.l.google.com.>,
  @preference=40,
  @ttl=943>]

address1.instance_variable_get(:@mx_servers_cache).object_id
=> 434860
address2.instance_variable_get(:@mx_servers_cache).object_id
=> 440840
```

**Comment by phoet** [comment]
😆 caching is hard. would you mind opening up a related issue?

**Comment by ianbayne** [comment]
Great catch, @natematykiewicz!

**Pull request merged**